### PR TITLE
Add types to utils.py

### PR DESCRIPTION
**Pull request by atgctg**
I added types to `utils.py`. Please review!

In case I add more, should I do them under separate PRs or under one big one?

(#250)

### CODE REVIEW

File: optax/_src/utils.py [hunk 1]
```diff
@@ -23,28 +25,29 @@
...
 
-def tile_second_to_last_dim(a):
+def tile_second_to_last_dim(a: chex.Array) -> chex.Array:
   ones = jnp.ones_like(a)
   a = jnp.expand_dims(a, axis=-1)
   return jnp.expand_dims(ones, axis=-2) * a
 
 
-def canonicalize_dtype(dtype):
+def canonicalize_dtype(dtype: Optional[jnp.dtype]) -> Optional[jnp.dtype]:
```
Comment (mkunesch) [inline optax/_src/utils.py:34]: I think we should make this `chex.ArrayDType` - I'm not sure whether `jnp.dtype` works if a user calls the function with e.g. just float. Also, that way we use chex wherever possible (consistent with `chex.Array`).

File: optax/_src/utils.py [hunk 2]
```diff
@@ -23,28 +25,29 @@
...
   """Canonicalise a dtype, skip if None."""
   if dtype is not None:
     return jax.dtypes.canonicalize_dtype(dtype)
   return dtype
 
 
-def cast_tree(tree, dtype):
+def cast_tree(tree: chex.ArrayTree,
+              dtype: Optional[jnp.dtype]) -> chex.ArrayTree:
```
Comment (mkunesch) [inline optax/_src/utils.py:42]: (as above, let's use `chex.ArrayDType`)

File: optax/_src/utils.py [hunk 3]
```diff
@@ -67,52 +70,58 @@ def set_diags(a, new_diags):
...
+  def __init__(self, loc: chex.Array, log_scale: chex.Array):
     self._log_scale = log_scale
     self._scale = jnp.exp(log_scale)
     self._mean = loc
     self._param_shape = jax.lax.broadcast_shapes(
         self._mean.shape, self._scale.shape)
 
-  def sample(self, shape, seed):
+  def sample(self, shape: Sequence[int],
+             seed: jax.random.PRNGKey) -> chex.Array:
```
Comment (mkunesch) [inline optax/_src/utils.py:84]: Let's use `chex.PRNGKey` for consistency with using chex for arrays.

File: optax/_src/utils.py [hunk 4]
```diff
@@ -67,52 +70,58 @@ def set_diags(a, new_diags):
...
     return jnp.sum(log_prob, axis=sum_axis)
 
   @property
-  def log_scale(self):
+  def log_scale(self) -> chex.Array:
     return self._log_scale
 
   @property
-  def params(self):
+  def params(self) -> List[chex.Array]:
```
Comment (mkunesch) [inline optax/_src/utils.py:102]: I think we should prefer the abstract type Sequence in this case - just so that we don't make any promises where they aren't strictly necessary. We might want to return a tuple or named tuple in the future.

**Review by mkunesch — CHANGES_REQUESTED** [review]
Thank you very much for the contribution! This looks great!

I've just added minor comments on using chex wherever possible and an abstract return type.

File: optax/_src/utils.py [hunk 5]
```diff
@@ -67,52 +70,58 @@ def set_diags(a, new_diags):
...
 
 
+def multi_normal(loc: chex.Array,
+                 log_scale: chex.Array) -> MultiNormalDiagFromLogScale:
+  return MultiNormalDiagFromLogScale(loc=loc, log_scale=log_scale)
+
+
 @jax.custom_vjp
-def _scale_gradient(inputs: chex.Array, scale: float)  -> chex.ArrayTree:
+def _scale_gradient(inputs: chex.Array, scale: float) -> chex.ArrayTree:
```
Comment (mkunesch) [inline optax/_src/utils.py:112]: Ah sorry, I think these inputs should be `chex.ArrayTree` as the gradients could be a nested tree of arrays.

**Review by mkunesch — CHANGES_REQUESTED** [review]
(thanks for the pointing out `chex.Array` vs `chex.ArrayTree`. I just checked this and there is one function where it should be `ArrayTree` instead of `Array`)

**Comment by mtthss** [comment]
There is still one change requested by @mkunesch, 
@atgctg could you address it so we get this submitted?

**Comment by mtthss** [comment]
This branch has conflicts that must be resolved

**Comment by atgctg** [comment]
Resolved it

**Review by mkunesch — APPROVED** [review]
Great, thanks a lot! Looks great to me!

**Comment by mkunesch** [comment]
Hi,

just to give a brief update: the delay in merging is due to internal tests having raised a bug: the + operator in line 85 does not work with the type annotation `Sequence[int]`. I'd propose to just cast shape to `tuple` in line 85. If that's okay with you I can add this small change in during merging (it would still be shown as your commits of course).

**Comment by atgctg** [comment]
Hi, sorry for not catching that. 

Please go ahead with your fix and thanks for letting me know!
